### ISSUE TITLE
Project 3 Proposal: Partial Redundancy Elimination

### ISSUE THREAD

**Issue by Neroysq** [sampsyo/cs6120#105]
**What will you do?**
I will implement a partial redundancy elimination pass. The goal is to eliminate expressions that are redundant on some but not necessarily all paths through a program.

**How will you do it?**
I will implement [lazy code motion](https://dl.acm.org/citation.cfm?id=143136).
It should be similar to #60 once I finish normalizing expressions.

**How will you empirically measure success?**
The outline for the evaluation would be similar to #60. I will first test the correctness by check:
* no non-safe expressions are eliminated, such as insertions that can cause an exception.
* no expressions are eliminated by introducing new computations in any path.
* the result after this optimization matches the original.

Then for performance, I will count
* lines of code
* computations executed 

and compare it to the original code.

I might also compare it to LLVM's `-gvnpre` pass.

The big difference from #60 would be that I will investigate existing benchmarks for LLVM/C++/C, and hopefully produce a more realistic and full-scale evaluation section.

**Team members:**
me.

**Comment by sampsyo** [comment]
Sounds pretty good to me! I know this advice will be hard to follow, but I'm going to offer it anyway:

Please try building the simplest possible version of lazy code motion that requires the minimum implementation effort. Then, spend most of your time doing a thorough evaluation to compare to `-gvnpre` in the built-in LLVM optimizations on an existing, full-scale benchmark suite. It's 100% OK (and totally expected) for your pass to do worse than the LLVM one—but investigating *why* this is and *how much* slower it is will be interesting!
